ARCHITECTURE.md

```diff
@@ -0,0 +1,226 @@
+# Istio operator architecture
+
+## Introduction
+
+The operator code is divided roughly into five areas:
+
+1. [New API](The_new_API) and related infrastructure. The new API is expressed as a
+[proto](https://github.com/istio/operator/blob/master/pkg/apis/istio/v1alpha2/istiocontrolplane_types.proto) and
+compiled to [Go structs](https://github.com/istio/operator/blob/master/pkg/apis/istio/v1alpha2/istiocontrolplane_types.pb.go). The new
+API has pass-through fields to the Helm values.yaml API, but these are additionally validated through a
+[schema](https://github.com/istio/operator/blob/master/pkg/apis/istio/v1alpha2/values_types.go).
+1. [Controller](K8s_controller) code. The code comprises the K8s listener, webhook and logic for reconciling the cluster
+to an IstioControlPlane CR.
+1. [Manifest creation](Manifest_creation) code. Manifest rendering is done by overlaying user settings on top of a
+selected profile and passing the resulting values to a renderer in the Helm library to create manifests. Further
+overlays on the created manifests may be applied if they are defined in the user settings.
+1. [CLI](CLI) code. CLI code shares the IstioControlPlane API with the controller, but allows manifests to be generated
+and optionally applied from the command line without the need to run a privileged controller in the cluster.
+1. [Migration tools](Migration_tools). The migration tools are intended to automate configuration migration from Helm to
+the operator.
+
+The operator code uses the new Helm charts in the [istio/installer](https://github.com/istio/installer) repo. It is not
+compatible with the older charts in [istio/istio](https://github.com/istio/istio/tree/master/install/kubernetes/helm).
+See the istio/installer repo for details about the new charts and why they were created. Briefly, the new charts
+are intended to support more realistic production deployments of Istio and hitless upgrades with canarying.
+
+## Terminology
+
+Throughout the document, the following terms are used:
+
+- New API: The API directly defined in the
```

Comment:
"New API" won't age well :) Can we use `IstioControlPlane` instead? We can add a version qualifier if necessary to different the API maturity.

Reply:
Good point, done. I've changed references to new and old APIs throughout the doc.